python/cvmfs/repository.py

```diff
@@ -198,38 +205,223 @@ def _pop_catalog(self):
+        Fetcher.__init__(self, local_repo, cache_dir)
+
+
+    def _retrieve_file(self, file_name):
+        full_path = self._make_file_uri(file_name)
+        if os.path.exists(full_path):
+            compressed_file = open(full_path, 'r')
+            decompressed_content = zlib.decompress(compressed_file.read())
+            compressed_file.close()
+            cached_file = self.cache.add(file_name)
+            self._write_content_into_file(decompressed_content, cached_file)
+            return self.cache.get(file_name)
+        else:
+            raise FileNotFoundInRepository(file_name)
+
+
+    def retrieve_raw_file(self, file_name):
+        """ Retrieves the file directly from the source """
+        full_path = self._make_file_uri(file_name)
+        if os.path.exists(full_path):
+            cached_file = self.cache.add(file_name)
+            compressed_file = open(full_path, 'r')
+            Fetcher._write_content_into_file(compressed_file.read(), cached_file)
+            compressed_file.seek(0)
+            return compressed_file
+        raise FileNotFoundInRepository(file_name)
+
+
+
+class RemoteFetcher(Fetcher):
+    """ Retrieves files from the local cache if found, and from
+    remote otherwise
+    """
+
+    def __init__(self, repo_url, cache_dir=''):
+        Fetcher.__init__(self, repo_url, cache_dir)
+
+    @staticmethod
+    def _download_content_and_store(cached_file, file_url):
+        response = requests.get(file_url, stream=True)
+        if response.status_code != requests.codes.ok:
+            raise FileNotFoundInRepository(file_url)
+        for chunk in response.iter_content(chunk_size=4096):
+            if chunk:
+                cached_file.write(chunk)
+        cached_file.close()
+
+
+    @staticmethod
+    def _download_content_and_decompress(cached_file, file_url):
+        response = requests.get(file_url, stream=False)
+        if response.status_code != requests.codes.ok:
+            raise FileNotFoundInRepository(file_url)
+        decompressed_content = zlib.decompress(response.content)
+        Fetcher._write_content_into_file(decompressed_content, cached_file)
+
+
+    def _retrieve_file(self, file_name):
+        file_url = self._make_file_uri(file_name)
+        cached_file = self.cache.add(file_name)
+        self._download_content_and_decompress(cached_file, file_url)
+        return self.cache.get(file_name)
+
+
+    def retrieve_raw_file(self, file_name):
+        cached_file = self.cache.add(file_name)
+        file_url = self._make_file_uri(file_name)
+        self._download_content_and_store(cached_file, file_url)
+        return self.cache.get(file_name)
 
-    __metaclass__ = abc.ABCMeta
 
-    def __init__(self):
+
+class Repository:
+    """ Wrapper around a CVMFS Repository representation """
+
+    def __init__(self, source, cache_dir=''):
+        if source == '':
+            raise Exception('source cannot be empty')
+        self.__determine_source(source, cache_dir)
```

Comment:
Totally true, it feels weird when the fetcher attribute appear from nowhere. Fixed! ;)